File: app/src/main/java/com/numero/colorpalette/activity/MainActivity.java
package com.numero.colorpalette.activity;

import android.graphics.Color;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.TabLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import android.support.v4.view.ViewPager;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import com.numero.colorpalette.R;
import com.numero.colorpalette.model.ColorStyle;
import com.numero.colorpalette.model.color.MaterialColor;
import com.numero.colorpalette.view.adapter.PagerAdapter;

import io.reactivex.Observable;

public class MainActivity extends AppCompatActivity implements ViewPager.OnPageChangeListener {

    private Toolbar toolbar;
    private AppBarLayout appBarLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        appBarLayout = findViewById(R.id.appbar);

        PagerAdapter pagerAdapter = new PagerAdapter(getApplicationContext(), getSupportFragmentManager());

        ViewPager viewPager = findViewById(R.id.container);
        viewPager.setAdapter(pagerAdapter);
        viewPager.addOnPageChangeListener(this);

        TabLayout tabLayout = findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);

        MaterialColor color = ColorStyle.values()[0].getColor();
        setToolbarColor(color);
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_about_app:
                startActivity(AboutActivity.createIntent(this));
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
        if (position >= ColorStyle.values().length) {
            return;
        }
        MaterialColor color = ColorStyle.values()[position].getColor();
        setToolbarColor(color);
    }

    @Override
    public void onPageScrollStateChanged(int state) {
    }

    private void setToolbarColor(@NonNull MaterialColor materialColor) {
        Observable.just(materialColor)
                .map(MaterialColor::getColor500)
                .map(Color::parseColor)
                .subscribe(color -> {
                    toolbar.setBackgroundColor(color);
                    appBarLayout.setBackgroundColor(color);
                });
        Observable.just(materialColor)
                .map(MaterialColor::getColor700)
                .map(Color::parseColor)
                .subscribe(color -> {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        getWindow().setStatusBarColor(color);
                    }
                });
    }
}

File: app/src/main/java/com/numero/colorpalette/activity/MainActivity.kt
package com.numero.colorpalette.activity

import android.content.ClipData
import android.content.ClipDescription
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Color
import android.os.Build
import android.support.v7.app.AppCompatActivity

import android.support.v4.view.ViewPager
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast

import com.numero.colorpalette.R
import com.numero.colorpalette.fragment.ColorListFragment
import com.numero.colorpalette.model.ColorStyle
import com.numero.colorpalette.model.color.MaterialColor
import com.numero.colorpalette.view.adapter.PagerAdapter
import io.reactivex.Observable
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), ViewPager.OnPageChangeListener, ColorListFragment.ColorListFragmentListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)

        val pagerAdapter = PagerAdapter(applicationContext, supportFragmentManager)

        viewPager.adapter = pagerAdapter
        viewPager.addOnPageChangeListener(this)

        tabs.setupWithViewPager(viewPager)

        val color = ColorStyle.values()[0].color
        setToolbarColor(color)
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_about_app -> {
                startActivity(AboutActivity.createIntent(this))
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

    override fun onPageSelected(position: Int) {
        if (position >= ColorStyle.values().size) {
            return
        }
        val color = ColorStyle.values()[position].color
        setToolbarColor(color)
    }

    override fun onPageScrollStateChanged(state: Int) {}

    private fun setToolbarColor(materialColor: MaterialColor) {
        Observable.just(materialColor)
                .map { it.color500 }
                .map { Color.parseColor(it) }
                .subscribe { color ->
                    toolbar!!.setBackgroundColor(color!!)
                    appbar!!.setBackgroundColor(color)
                }
        Observable.just(materialColor)
                .map { it.color700 }
                .map { Color.parseColor(it) }
                .subscribe { color ->
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        window.statusBarColor = color!!
                    }
                }
    }

    override fun onClickColor(color: String?) {
        if (color == null) {
            return
        }
        copyText(color)
        Toast.makeText(this, String.format("Copied color : %s", color), Toast.LENGTH_SHORT).show()
    }

    private fun copyText(text: String) {
        val clipboardManager = getSystemService(Context.CLIPBOARD_SERVICE)
        if (clipboardManager is ClipboardManager) {
            clipboardManager.primaryClip = ClipData(ClipDescription("copied color", arrayOf(ClipDescription.MIMETYPE_TEXT_URILIST)), ClipData.Item(text))
        }
    }
}
